### DISCUSSION TITLE
Error when debugging using `codelldb` EACCESS: permission denied

### DISCUSSION THREAD

**Discussion by BogdanFloris** [mrcjkb/rustaceanvim#217] [Q&A] [answered]
### Neovim version (nvim -v)

v0.9.0

### Operating system/version

MacOS 14.2.1

### Output of :checkhealth rustaceanvim

```console
==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- OK [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap) installed.

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 0.3.1831-standalone (39ad79bec 2024-02-04)
- OK Cargo: found cargo 1.74.0 (ecb9851af 2023-10-18)
- OK rustc: found rustc 1.74.0 (79e9716c9 2023-11-13)
- OK debug adapter: found codelldb 

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

### How to reproduce the issue

```console
Open any Rust project, hover over `main` and click on Debug.
```

### Expected behaviour

The DAP UI debugger opens

### Actual behaviour

The following error appears in messages:
```
.local/share/nvim/lazy/nvim-dap/lua/dap/session.lua:1186: EACCES: permission denied
```

I also set the DAP log level to DEBUG and I got the following log:
```
[ DEBUG ] 2024-02-07T19:41:03Z+0200 ] ...oris/.local/share/nvim/lazy/nvim-dap/lua/dap/session.lua:1165 ]	"Starting debug adapter server executable"	{
  args = { "--port", "64778" },
  command = ""
}
```

### The minimal config used to reproduce this issue.

This is my rustaceanvim config using `lazy`:

```Lua

return {
  'mrcjkb/rustaceanvim',
  version = '^4',
  ft = { 'rust' },
  opts = {
    server = {
      on_attach = on_attach,
      standalone = true,
      settings = {
        ['rust-analyzer'] = {
          cargo = {
            allFeatures = true,
            loadOutDirsFromCheck = true,
            runBuildScripts = true,
          },
          -- Add clippy lints for Rust.
          checkOnSave = {
            allFeatures = true,
            command = 'clippy',
            extraArgs = { '--', '-W', 'clippy::pedantic' },
          },
          procMacro = {
            enable = true,
            ignored = {
              ['async-trait'] = { 'async_trait' },
              ['napi-derive'] = { 'napi' },
              ['async-recursion'] = { 'async_recursion' },
            },
          },
        },
      },
    },
  },
  config = function(_, opts)
    vim.g.rustaceanvim = vim.tbl_deep_extend('force', {}, opts or {})
  end,
}
```

**Comment by mrcjkb** [comment]
Hey :wave: 

I've converted this to a discussion, as the error is thrown by nvim-dap, and is unlikely to be caused by a bug in rustaceanvim.

Are you able to run `codelldb --port 64778` from a terminal?
It looks like your OS or firewall could be blocking the port that nvim-dap is telling codelldb to listen on.

**Reply by BogdanFloris** [reply]
👋🏻 Thanks for the quick answer! That makes sense turning this into a discussion. I wasn't sure if this is `ivim-dap` related or there was something wrong with my configuration.

I can run `codelldb --port 64778` just fine. (`codelldb` is installed by Mason so I run it from there).

**Reply by mrcjkb** [reply]
> codelldb is installed by Mason so I run it from there

:thinking: I wonder if there's an issue with getting the codelldb executable path from the mason registry on MaxOS?
It's also suspicious that `command` field in your log entry is an empty string (it should be `codelldb` or the path to your codelldb mason install).

If you're able to run it from the terminal, rustaceanvim shouldn't need to get the info from mason.
Could you please try the `codelldb-prioritise-path` branch, and see if that fixes it?

**Reply by BogdanFloris** [reply]
Just tried it, it's still the same issue with the same log from DAP:
```
[ DEBUG ] 2024-02-07T20:47:37Z+0200 ] ...oris/.local/share/nvim/lazy/nvim-dap/lua/dap/session.lua:1165 ]	"Starting debug adapter server executable"	{
  args = { "--port", "65083" },
  command = ""
}
```
I'll try setting the command directly in DAP and see if it works.

**Reply by mrcjkb** [reply]
Another thing you could try:
Does the same error occur when you debug using the `:RustLsp debuggables` command?

**Reply by BogdanFloris** [reply]
Yeah, the `debuggables` commands result in the same error.

I don't think it's related to the executable command not being correctly set. I tried setting it directly in the `sessions.lua` file in `dap` and it's still resulting in the same issue. Very weird stuff.

```
[ DEBUG ] 2024-02-07T21:12:06Z+0200 ] ...oris/.local/share/nvim/lazy/nvim-dap/lua/dap/session.lua:1166 ]	"Starting debug adapter server executable"	{
  args = { "--port", "65214" },
  command = "/Users/bogdanfloris/.local/share/nvim/mason/packages/codelldb/codelldb"
}
```

**Reply by mrcjkb** [reply]
Hmm, maybe your firewall is blocking the codelldb process spawned by neovim?
Or maybe your OS is blocking neovim from spawning the codelldb process?

Or perhaps the executable installed by mason doesn't have exec permissions?
When you run it from the terminal, are you running the binary installed my mason?

**Reply by BogdanFloris** [reply, marked as answer]
It wasn't that. Apparently in MacOS, you need to manually setup the `codelldb` handler in `mason-nvim-dap`.

For anyone that stumbles upon this issue, here's the configuration that fixed it: https://github.com/BogdanFloris/dotfiles/blob/master/files/.config/nvim/lua/plugins/debug.lua
